### ISSUE TITLE
Got 'NoSuchBucket' when bucket has defined in localstack

### ISSUE THREAD

**Issue by YaYun-Wang** [localstack/localstack#8262]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

I am using localstack to mock aws service by docker-compose, and take 2 measures to make sure bucket test has been created:
1. execute command: sh -c "aws --endpoint-url=http://localstack:4566 s3 mb s3://test" in docker-compose.yml
2. Defined the data directory so that localstack can load s3 data when starting: DATA_DIR=/var/lib/localstack/data

But i got this error when i download my bucket through 's3Client.getObject('test', mykey)':
`com.amazonaws.services.s3.model.AmazonS3Exception: The specified bucket does not exist (Service: Amazon S3; Status Code: 404; Error Code: NoSuchBucket; Request ID: 045228a8-8ee7-4db2-a193-0e72c9724a8f; S3 Extended Request ID: null; Proxy: null)`

### Expected Behavior

Could download the bucket successfully through 's3Client.getObject('test', mykey)'

### How are you starting LocalStack?

With a docker-compose file

### Steps To Reproduce

Here is my docker-compose.yml:

version: '2.3'
services:
  localstack:
    image: localstack/localstack:latest
    ports:
      - 4566:4566
    command: sh -c "aws --endpoint-url=http://localstack:4566 s3 mb s3://test"
    environment:
      - SERVICES=s3,kms
      - DOCKER_HOST=unix:///var/run/docker.sock
      - PORT_WEB_UI=4567
      - START_WEB=1
      - DEBUG=1
      - DATA_DIR=/var/lib/localstack/data
      - DEFAULT_REGION=cn-north-1
      - AWS_ACCESS_KEY_ID=test
      - AWS_SECRET_ACCESS_KEY=test
      - LEGACY_PERSISTENCE=1
      - HOSTNAME_EXTERNAL=localstack
      - LEGACY_EDGE_PROXY=1
    volumes:
      - './stubs/s3:/var/lib/localstack'
      - '/var/run/docker.sock:/var/run/docker.sock'
    healthcheck:
      test: '(key_id=$$(awslocal kms create-key | python3 -c ''import json,sys;obj=json.load(sys.stdin);print(obj["KeyMetadata"]["KeyId"])'')
        && echo $$key_id
        && awslocal kms create-alias --alias-name $$ALIAS --target-key-id $$key_id) || true'
      interval: 5s
      timeout: 5s
      retries: 120

My directiry structure is:
--docker-compose.yml
--stubs
  --s3
    --data
       --recorded_api_calls.json

### Environment

```markdown
- OS: ubuntu1~20.04.1
- LocalStack:The latest version
```

**Comment by bentsku** [comment]
Hi @YaYun-Wang and thanks for your report!

From your docker-compose file, I can see there are a lot of deprecated variables non compatible with the latest version of LocalStack (`DATA_DIR`, 'LEGACY_EDGE_PROXY`, `LEGACY_PERSISTENCE` mostly), some of them deprecated since LocalStack v1. 

The best way would be to start fresh with the new Docker Compose file we have in our repo: https://github.com/localstack/localstack/blob/master/docker-compose.yml

To create an initial state in your LocalStack container, you should make use of our new initialisation hooks, with the following documentation here: https://docs.localstack.cloud/references/init-hooks
There's even a example covering your use case of creating an S3 bucket: https://docs.localstack.cloud/references/init-hooks/#usage-example
And you can also see there how to mount the initialization bash script to create the bucket.

I hope this helps, don't hesitate if there's still anything.

**Comment by YaYun-Wang** [comment]
Thanks @bentsku , i got the same error 'NoSuchBucket' after i modify my docker-compose.yml refer to above link, here is my docker-compose.yml and init-aws.sh:

**docker-compose.yml**
version: '3.8'
services:
  localstack:
    image: localstack/localstack
    ports:
      - "4566:4566"
    environment:
      - DEBUG=1
      - DOCKER_HOST=unix:///var/run/docker.sock
    volumes:
      - "./localstack/init-aws.sh:/etc/localstack/init/ready.d/init-aws.sh" 
      - './localstack:/var/lib/localstack'
      - "/var/run/docker.sock:/var/run/docker.sock"

**init-aws.sh**:
#!/bin/bash
awslocal s3 mb s3://testBucket

**Directory structure is:**
--localstack
----init-aws.sh
--docker-compose.yml

And i noticed that there are some error logs indicate the init script was not executed:
`[  MainThread] localstack.runtime.init    : Init scripts discovered: {BOOT: [], START: [], READY: [], SHUTDOWN: []}`

Did i miss anything?

**Comment by jbosse** [comment]
I have the same issue. I am mapping the volume as ` "/etc/localstack/init_ready/:/etc/localstack/init/ready.d/"` and also see `Init scripts discovered: {BOOT: [], START: [], READY: [], SHUTDOWN: []}`

**Comment by bentsku** [comment]
@thrau, do you have an idea of what the issue could be?

**Comment by alexrashed** [comment]
Hi @jbosse!
I'll chime in here, because I think the root cause of your issues is the volume configuration.

> I have the same issue. I am mapping the volume as "/etc/localstack/init_ready/s3_startup.sh:/etc/localstack/init/ready.d/s3_startup.sh" and also see Init scripts discovered: {BOOT: [], START: [], READY: [], SHUTDOWN: []}

This actually makes sense, because of your volume mount: `/etc/localstack/init_ready/s3_startup.sh:/etc/localstack/init/ready.d/s3_startup.sh`
This means that you are mounting your _local_ directory `/etc/localstack/init_ready/s3_startup.sh` (mind the leading slash, making it an absolute path from your root) in the LocalStack container.
Since I expect that this file does not exist on your machine, Docker just mounts an empty directory at this place in the container. This explains why the `ls` command shows the "script", which in fact is actually a directory (which you can check with an `ls -al`).
In your `Dockerfile`, you are using the correct, relatively referenced file (it also wouldn't work otherwise due to the context boundary).
So your issue will be fixed when you change the absolute to a relative path.
Also, please make sure that your script has a shebang (f.e. `#!/bin/bash`) on top and has the executable flag set.

TL/DR:
```diff
   version: '3.8'
   services:
     localstack:
       ...
     volumes:
---    - "/etc/localstack/init_ready/s3_startup.sh:/etc/localstack/init/ready.d/s3_startup.sh"
+++    - "./etc/localstack/init_ready/s3_startup.sh:/etc/localstack/init/ready.d/s3_startup.sh"
       ...
```

I hope this resolves your issue, please let us know. Thanks!
